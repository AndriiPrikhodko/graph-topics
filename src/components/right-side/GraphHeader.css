.loaderIcon {
    margin-top: 3px;
    margin-left: 5px;
    animation: spin 2s infinite linear;
}

.save {
    margin-left: 4px;
    padding-left: 5px;
    padding-right: 5px;
    margin-top: 0;
}

#save-filename-input {
    background-color: #ccc;
    color: #111;
    border: none;
    font-size: 14pt;
    border-radius: 4px;
}

#save-filename-input:focus {
    outline: solid;
    outline-color: #111;
}

.clean {
    margin-left: auto;
}

.graph-header {
    padding-bottom: 5px;
    padding-left: 5px;
    max-width: 95%;
    display: flex;
}

@keyframes spin {
    from {
        transform: scale(1) rotate(30deg);
    }
    to {
        transform: scale(1) rotate(360deg);
    }
}