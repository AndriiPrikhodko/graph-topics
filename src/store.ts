import { configureStore } from '@reduxjs/toolkit'
import rootReducer from './reducers'
import thunk from 'redux-thunk'

const store = configureStore({
  reducer: {
    graphData: rootReducer.graphSlice,
    graphList: rootReducer.graphListSlice
  },
  middleware: [thunk],
})

export default store;