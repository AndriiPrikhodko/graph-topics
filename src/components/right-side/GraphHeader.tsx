import { graphAPI } from "../../api/graph.api"
import { TbFidgetSpinner } from 'react-icons/tb';
import { ImSpinner2 } from "react-icons/im"
import { ChangeEvent, useState } from "react"
import { useSelector, useDispatch } from 'react-redux'
import { loadGraphList } from "../../actions/graph-api.action";
import './GraphHeader.css'

const GraphHeader: React.FC = () => {
    const [graphLoading, setGraphLoading] = useState<boolean>(false)
    const [graphName, setGraphName] = useState('')
    const graphData = useSelector((store: Store) => store.graphData)
    const dispatch = useDispatch()

    const handleSaveGraphClick = () => {
        setGraphLoading(true)
        if(graphName.trim().length > 0){
            const payload:IPostGraphPayload = {
                name: graphName.trim(),
                data: graphData
            }

            graphAPI
            .save(payload)
            .then(() => dispatch(loadGraphList()))
            .catch(err => console.error(err))
            // .then(() => setGraphName(''))
            .finally(() =>
                setGraphLoading(false)
            )
        }
        else
            console.log('Enter graph name to save')
    }

    const handleInputNameChange = (event: ChangeEvent) => {
        const target = event.target as HTMLInputElement
        setGraphName(target.value)
    }

    return (
        <div className="graph-header">
            <input value={graphName} onChange={handleInputNameChange} placeholder="enter graph name"></input>
                <button className="save unselectable" onClick={handleSaveGraphClick}>Save</button>
            {(() => {
                if (graphLoading) {
                    return <ImSpinner2 className="loaderIcon" />
                }
            })()}
                
        </div>
    )
}

export default GraphHeader