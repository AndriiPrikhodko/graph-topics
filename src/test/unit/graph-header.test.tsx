import {  test } from 'vitest'
import { render } from '@testing-library/react';
import GraphHeader from '../../components/right-side/GraphHeader';
import store from '../../store';
import { Provider } from 'react-redux';

test('header', () => {
  const { getByTestId } = render(
    <Provider store={store}>
    <GraphHeader />
    </Provider>
  )
  const graphHeaderEl = getByTestId('graph-view-header')
  expect(graphHeaderEl).toMatchSnapshot()
})
