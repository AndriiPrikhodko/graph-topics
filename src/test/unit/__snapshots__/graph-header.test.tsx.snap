// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`header 1`] = `
<div
  class="graph-header"
  data-testid="graph-view-header"
>
  <div
    class="history-wrapper header-item"
  >
    <div
      class="icon-encircle disabled"
    >
      <svg
        class="action-icon"
        color="black"
        data-testid="icon-undo"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="color: black;"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          undo action
        </title>
        <path
          d="M464 440l-28.12-32.11c-22.48-25.65-43.33-45.45-72.08-58.7-26.61-12.26-60-18.65-104.27-19.84V432L48 252 259.53 72v103.21c72.88 3 127.18 27.08 161.56 71.75C449.56 284 464 335.19 464 399.26z"
        />
      </svg>
    </div>
    <div
      class="icon-encircle disabled"
    >
      <svg
        class="action-icon"
        color="black"
        data-testid="icon-redo"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="color: black;"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          redo action
        </title>
        <path
          d="M48 399.26C48 335.19 62.44 284 90.91 247c34.38-44.67 88.68-68.77 161.56-71.75V72L464 252 252.47 432V329.35c-44.25 1.19-77.66 7.58-104.27 19.84-28.75 13.25-49.6 33.05-72.08 58.7L48 440z"
        />
      </svg>
    </div>
  </div>
  <input
    class="header-item"
    data-testid="save-input"
    id="save-filename-input"
    placeholder="enter graph name"
    value=""
  />
  <button
    class="standard-button"
    data-testid="save-button"
  >
    Save
  </button>
  <div
    class="action-wrapper-header"
  >
    <div
      class="icon-encircle "
    >
      <svg
        class="action-icon"
        color="black"
        data-testid="icon-addGraphEdge"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="color: black;"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          link nodes
        </title>
        <path
          d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
        />
      </svg>
    </div>
    <div
      class="icon-encircle "
    >
      <svg
        class="action-icon"
        color="black"
        data-testid="icon-removeGraphEdge"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: black;"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          unlink nodes
        </title>
        <path
          d="M0 0h24v24H0z"
          fill="none"
          stroke="none"
        />
        <path
          d="M9 15l3 -3m2 -2l1 -1"
        />
        <path
          d="M11 6l.463 -.536a5 5 0 0 1 7.071 7.072l-.534 .464"
        />
        <path
          d="M3 3l18 18"
        />
        <path
          d="M13 18l-.397 .534a5.068 5.068 0 0 1 -7.127 0a4.972 4.972 0 0 1 0 -7.071l.524 -.463"
        />
      </svg>
    </div>
    <div
      class="icon-encircle "
    >
      <svg
        class="action-icon"
        color="black"
        data-testid="icon-deleteGraphNode"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="color: black;"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          delete node
        </title>
        <path
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
        />
      </svg>
    </div>
    <div
      class="icon-encircle "
    >
      <svg
        class="action-icon"
        color="darkorange"
        data-testid="icon-cancelIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="color: darkorange;"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          cancel action
        </title>
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm288.5 682.8L277.7 224C258 240 240 258 224 277.7l522.8 522.8C682.8 852.7 601 884 512 884c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372c0 89-31.3 170.8-83.5 234.8z"
          fill="#333"
        />
        <path
          d="M512 140c-205.4 0-372 166.6-372 372s166.6 372 372 372c89 0 170.8-31.3 234.8-83.5L224 277.7c16-19.7 34-37.7 53.7-53.7l522.8 522.8C852.7 682.8 884 601 884 512c0-205.4-166.6-372-372-372z"
          fill="#E6E6E6"
        />
      </svg>
    </div>
  </div>
  <div
    class="right"
  >
    <button
      class="standard-button"
      data-testid="clear-button"
    >
      Clear
    </button>
  </div>
</div>
`;
